synthesizer/program/src/logic/command/contains.rs

```diff
@@ -84,9 +84,9 @@ impl<N: Network> Contains<N> {
             CallOperator::Resource(mapping_name) => (*stack.program_id(), mapping_name),
         };
 
-        // Ensure the mapping exists in storage.
-        if !store.contains_mapping_confirmed(&program_id, &mapping_name)? {
-            bail!("Mapping '{program_id}/{mapping_name}' does not exist in storage");
+        // Ensure the mapping exists.
+        if !store.contains_mapping_speculative(&program_id, &mapping_name)? {
```

Comment:
I assume the intention for this change is to allow constructors to be properly executed. How can we check that we hit all of these cases/opcodes?

Reply:
Also, does this now enable transactions to be executed in the same block that their program was deployed? (via a malicious validator injecting the execution after the deployment into his proposal)

Reply:
We could add an integration test that uses all of the opcodes.
The current one does a subset of them.

Reply:
> Also, does this now enable transactions to be executed in the same block that their program was deployed? (via a malicious validator injecting the execution after the deployment into his proposal).

I don't think so because all transactions are verified in `prepare_for_speculate` before any (speculative) deployment happens.

Reply:
@ljedrz during review, specific attention was called to this change. Do you see any issues or potential vulnerabilities?

Reply:
This change means that it would now suffice for the mapping to be part of the ongoing atomic operation, i.e. it doesn't have to be _physically_ in the persistent storage yet. As long as there is no way of undoing/changing that mapping throughout the rest of this specific storage operation, this should technically be fine, since at the end of it, both the mapping and all the remaining storage entries will have been persisted.

Reply:
In the atomic operation (for constructors), the mapping can have entries added or removed. If the execution of the constructor fails, then then entire batch including the initialization of the mappings should be aborted.

Reply:
this is how it works for all the atomic operations, and this change doesn't affect it

Reply:
We also added the restriction that a program cannot be executed in the same block that it was upgraded in.
In this case, the execution will get aborted.